File: python/ocp-solid.py
# PYTHON
# The Open/Closed Principle states 
# that a class should be open for extension 
# but closed for modification. This means 
# that we should be able to add new 
# functionality without modifying existing 
# code, which we achieve using inheritance 
# and polymorphism.

# Example of clean SOLID code:
import math

# Base Shape class
class Shape:
    def __init__(self, color="black"):
        self.color = color  # All shapes have a color

    def draw(self):
        """Default draw method to be overridden in subclasses."""
        print(f"Drawing a shape with color: {self.color}")

    def change_color(self, new_color):
        """Method to change the color of the shape."""
        self.color = new_color
        print(f"The color of the shape has been changed to: {self.color}")

# Circle class extends Shape
class Circle(Shape):
    def __init__(self, radius, color="black"):
        super().__init__(color)  # Call the parent constructor to set the color
        self.radius = radius  # Circle-specific property: radius

    def draw(self):
        """Circle-specific draw method"""
        print(f"Drawing a circle with radius {self.radius} and color: {self.color}")
        print(f"Area of circle: {math.pi * self.radius ** 2}")

    def resize(self, new_radius):
        """Method to resize the circle by changing the radius"""
        self.radius = new_radius
        print(f"The radius of the circle has been changed to: {self.radius}")

# Rectangle class extends Shape
class Rectangle(Shape):
    def __init__(self, width, height, color="black"):
        super().__init__(color)  # Call the parent constructor to set the color
        self.width = width  # Rectangle-specific property: width
        self.height = height  # Rectangle-specific property: height

    def draw(self):
        """Rectangle-specific draw method"""
        print(f"Drawing a rectangle with width {self.width}, height {self.height}, and color: {self.color}")
        print(f"Area of rectangle: {self.width * self.height}")

    def resize(self, new_width, new_height):
        """Method to resize the rectangle by changing width and height"""
        self.width = new_width
        self.height = new_height
        print(f"The dimensions of the rectangle have been changed to: {self.width}x{self.height}")

# Square class extends Rectangle
class Square(Rectangle):
    def __init__(self, side_length, color="black"):
        # A square is a special case of a rectangle where width and height are equal
        super().__init__(side_length, side_length, color)

    def draw(self):
        """Square-specific draw method"""
        print(f"Drawing a square with side length {self.width} and color: {self.color}")
        print(f"Area of square: {self.width ** 2}")

    def resize(self, new_side_length):
        """Method to resize the square by changing side length"""
        self.width = new_side_length
        self.height = new_side_length
        print(f"The side length of the square has been changed to: {self.width}")

# Usage

# Create a Circle object, draw it, and modify its properties
circle = Circle(5, "red")
circle.draw()
circle.resize(10)
circle.draw()
circle.change_color("blue")
circle.draw()

# Create a Rectangle object, draw it, and modify its properties
rectangle = Rectangle(4, 6, "blue")
rectangle.draw()
rectangle.resize(8, 10)
rectangle.draw()

# Create a Square object, draw it, and modify its properties
square = Square(4, "green")
square.draw()
square.resize(6)
square.draw()
square.change_color("yellow")
square.draw()

File: javascript/openClosed/ocp-solid.js
//Open / Close
//Open to extention, closed to change or modification
//Needs to be able to be extend functionality without modifying code
//Achieved through polymorphism (e.g., inheritance, interfaces, or composition).

// GOOD because class circle can't be used for another class like square, 
// nor rectangle or mix match so we put them in 
// there own reusable class and extend them that way
//No need to modify the code in the module when done properly
// Base Shape class
class Shape {
  //All shapes will have whats on this/constructor.
  constructor(color = "black") {
    this.color = color; // All shapes will have a color
  }

  // Default draw method (to be overwritten in subclass)
  draw() {
    console.log(`Drawing a shape with color: ${this.color}`);
  }

  // Method to change the color of the shape
  changeColor(newColor) {
    this.color = newColor;
    console.log(`The color of the shape has been changed to: ${this.color}`);
  }
}

// Circle class extends Shape
class Circle extends Shape {
  constructor(radius, color = "black") {
    super(color); // Call the parent constructor to set the color
    this.radius = radius; // Circle/class -specific property: radius
  }

  // Circle-specific draw method
  draw() {
    console.log(`Drawing a circle with radius ${this.radius} and color: ${this.color}`);
    console.log(`Area of circle: ${Math.PI * this.radius * this.radius}`);
  }

  // Method to resize the circle by changing the radius
  resize(newRadius) {
    this.radius = newRadius;
    console.log(`The radius of the circle has been changed to: ${this.radius}`);
  }
}

// Rectangle class extends Shape
class Rectangle extends Shape {
  constructor(width, height, color = "black") {
    super(color); // Call the parent constructor to set the color with super func
    this.width = width; // Rectangle-specific property: width
    this.height = height; // Rectangle-specific property: height
  }

  // Rectangle-specific draw method
  draw() {
    console.log(`Drawing a rectangle with width ${this.width}, height ${this.height}, and color: ${this.color}`);
    console.log(`Area of rectangle: ${this.width * this.height}`);
  }

  // Method to resize the rectangle by changing width and height
  resize(newWidth, newHeight) {
    this.width = newWidth;
    this.height = newHeight;
    console.log(`The dimensions of the rectangle have been changed to: ${this.width}x${this.height}`);
  }
}

// Square class extends Rectangle
// Square is just a special case of Rectangle where width and height are equal
class Square extends Rectangle {
  constructor(sideLength, color = "black") {
    // Call the parent constructor and overwrite with equal width and height
    super(sideLength, sideLength, color); 
  }

  // Square-specific draw method
  draw() {
    console.log(`Drawing a square with side length ${this.width} and color: ${this.color}`);
    console.log(`Area of square: ${this.width * this.width}`);
  }

  // Method to resize the square by changing side length
  resize(newSideLength) {
    this.width = newSideLength;
    this.height = newSideLength;
    console.log(`The side length of the square has been changed to: ${this.width}`);
  }
}

// Usage

// Create a circle, draw it, and modify its properties
const circle = new Circle(5, "red");
circle.draw(); 
// Output:
// Drawing a circle with radius 5 and color: red
// Area of circle: 78.53981633974483

circle.resize(10); 
// Output: 
// The radius of the circle has been changed to: 10
circle.draw(); 
// Output:
// Drawing a circle with radius 10 and color: red
// Area of circle: 314.1592653589793

circle.changeColor("blue");
// Output:
// The color of the shape has been changed to: blue
circle.draw(); 
// Output:
// Drawing a circle with radius 10 and color: blue
// Area of circle: 314.1592653589793

// Create a rectangle, draw it, and modify its properties
const rectangle = new Rectangle(4, 6, "blue");
rectangle.draw(); 
// Output:
// Drawing a rectangle with width 4, height 6, and color: blue
// Area of rectangle: 24

rectangle.resize(8, 10);
// Output:
// The dimensions of the rectangle have been changed to: 8x10
rectangle.draw(); 
// Output:
// Drawing a rectangle with width 8, height 10, and color: blue
// Area of rectangle: 80

// Create a square, draw it, and modify its properties
const square = new Square(4, "green");
square.draw(); 
// Output:
// Drawing a square with side length 4 and color: green
// Area of square: 16

square.resize(6);
// Output:
// The side length of the square has been changed to: 6
square.draw(); 
// Output:
// Drawing a square with side length 6 and color: green
// Area of square: 36

square.changeColor("yellow");
// Output:
// The color of the shape has been changed to: yellow
square.draw(); 
// Output:
// Drawing a square with side length 6 and color: yellow
// Area of square: 36


//BAD Example 
// JavaScript code:
//Requires modification
//Too many different functions in one class
//Using a string this way is not good for scaling
// class Shape {
//  draw(builtInShape) {
//   if(builtInShape === "circle") {
//     console.log("cirlcle");
//   } else if(builtInShape === "rectangle") {
//     console.log("rectangle");
//   } else if(builtInShape === "square") {
//     console.log("square")
//   } else {
//     console.log("unknown class")
//   }
//  }
// };

var iceCreamFlavor = ["choc", "vanil"];
var iceCreamMaker = {
  makeIceCream(flavor){
    if(iceCreamFlavors.indexOf(flavor) > -1){
      console.log("success")
    } else {
      console.log("failed ")
    }
  }

}